pkg/virt-controller/watch/clone/clone.go

```diff
@@ -83,193 +89,194 @@ func (ctrl *VMCloneController) execute(key string) error {
+			_, syncInfo = ctrl.createRestoreFromVm(vmClone, source, snapshot.Name, syncInfo)
+			return syncInfo
+		}
 
-			restore, syncInfo.err = ctrl.client.VirtualMachineRestore(restore.Namespace).Create(context.Background(), restore, v1.CreateOptions{})
-			if syncInfo.err != nil {
-				return addErrorToSyncInfo(syncInfo, fmt.Errorf("failed creating restore %s for clone %s: %v", restore.Name, vmClone.Name, syncInfo.err))
-			}
+		_, syncInfo = ctrl.verifyRestoreReady(vmClone, source.Namespace, syncInfo)
+		if syncInfo.toReenqueue() {
+			return syncInfo
+		}
 
-			ctrl.logAndRecord(vmClone, RestoreCreated, fmt.Sprintf("created restore %s for clone %s", restore.Name, vmClone.Name))
-			syncInfo.restoreName = restore.Name
+		fallthrough
+
+	case clonev1alpha1.CreatingTargetVM:
 
-			logger.V(defaultVerbosityLevel).Infof("restore %s was just created, reenqueuing to let snapshot time to finish", restore.Name)
+		syncInfo = ctrl.verifyVmReady(vmClone, syncInfo)
+		if syncInfo.toReenqueue() {
 			return syncInfo
 		}
 
-		// Make sure restore is ready for use
-		obj, exists, err := ctrl.restoreInformer.GetStore().GetByKey(getKey(*vmClone.Status.RestoreName, source.Namespace))
-		if !exists {
-			return addErrorToSyncInfo(syncInfo, fmt.Errorf("restore %s is not created yet for clone %s", *vmClone.Status.SnapshotName, vmClone.Name))
-		} else if err != nil {
-			return addErrorToSyncInfo(syncInfo, fmt.Errorf("error getting snapshot %s from cache for clone %s: %v", *vmClone.Status.SnapshotName, vmClone.Name, err))
+		syncInfo = ctrl.cleanupSnapshot(vmClone, syncInfo)
+		if syncInfo.toReenqueue() {
+			return syncInfo
 		}
 
-		restore := obj.(*snapshotv1alpha1.VirtualMachineRestore)
-		logger.Infof("found target restore %s for clone %s", restore.Name, vmClone.Name)
-
-		if virtsnapshot.VmRestoreProgressing(restore) {
-			logger.V(defaultVerbosityLevel).Infof("restore %s for clone %s is not ready to use yet", restore.Name, vmClone.Name)
+		syncInfo = ctrl.cleanupRestore(vmClone, syncInfo)
+		if syncInfo.toReenqueue() {
 			return syncInfo
 		}
 
-		ctrl.logAndRecord(vmClone, RestoreReady, fmt.Sprintf("restore %s for clone %s is ready to use", restore.Name, vmClone.Name))
-		syncInfo.restoreReady = true
-		syncInfo.targetVMName = restore.Spec.Target.Name
+	default:
+		log.Log.Object(vmClone).Infof("clone %s is in phase %s - nothing to do", vmClone.Name, string(vmClone.Status.Phase))
+	}
+
+	return syncInfo
+}
+
+func (ctrl *VMCloneController) syncSourceSnapshotTargetVM(source *snapshotv1alpha1.VirtualMachineSnapshot, vmClone *clonev1alpha1.VirtualMachineClone) syncInfoType {
+	syncInfo := syncInfoType{logger: log.Log.Object(vmClone)}
+
+	switch vmClone.Status.Phase {
+	case clonev1alpha1.PhaseUnset, clonev1alpha1.SnapshotInProgress:
+
+		syncInfo.snapshotName = source.Name
+		source, syncInfo = ctrl.verifySnapshotReady(vmClone, source.Name, source.Namespace, syncInfo)
+		if syncInfo.toReenqueue() {
+			return syncInfo
```

Comment:
I don't think this case is properly handled.  The controller will not reconcile when the snapshot becomes ready because the snapshot is not owned by the VMClone resource.

To handle this, you can requeue after X seconds or add a "vmsnapshot" index to VMClones that maps the source snapshot to the VMClone(s)

Reply:
Thanks!
Main related changes are in https://github.com/kubevirt/kubevirt/pull/8458/commits/332cdc122442578f81982a683fc190e7cd0bdf10

Reply:
As said before, this is already handled here: https://github.com/kubevirt/kubevirt/commit/332cdc122442578f81982a683fc190e7cd0bdf10. With this commit the clone controller won't only wake up to updates of snapshot it owns - but will also wake up if a snapshot that is the source of one of the clones updates.

> Same question, what about the case when snapshot is ready but the VMClone status phase shows its in progress?

As written above - when the snapshot will become ready, the controller would wake up and reenqueue the clone object.

Am I missing something here?